### PR TITLE
new functions for creation of empty VM 

### PR DESCRIPTION
**Pull request by vbauzysvmware**
New functions needed implementing empty VM functionality in terraform.

* Add empty MV
* Query all media is used to find media to attach as boot image
* UpdateVmSpecSectionAsync/UpdateVmSpecSection

### CODE REVIEW

File: govcd/media.go [hunk 1]
```diff
@@ -637,3 +637,42 @@ func (mediaRecord *MediaRecord) Delete() (Task, error) {
...
+
+// QueryAllMedia returns media image found in system using `name` as query.
+func (vdc *Vdc) QueryAllMedia(mediaName string) ([]*MediaRecord, error) {
+	util.Logger.Printf("[TRACE] Querying medias by name\n")
+
+	if mediaName == "" {
+		return nil, errors.New("media name is empty")
+	}
+
+	typeMedia := "media"
```
Comment (Didainius) [inline govcd/media.go]: Just curious - can there be a mix of media (admin and non-admin) when a user is admin?

File: types/v56/vm_types.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	MemoryResourceMb  *MemoryResourceMb `xml:"MemoryResourceMb"`            // Memory compute resources. This parameter may be omitted when using the VmSpec to update the contents of an existing VM.
+	MediaSection      *MediaSection     `xml:"MediaSection,omitempty"`      // The media devices of this VM.
+	DiskSection       *DiskSection      `xml:"DiskSection,omitempty"`       // virtual disks of this VM.
+	HardwareVersion   *HardwareVersion  `xml:"HardwareVersion"`             // vSphere name of Virtual Hardware Version of this VM. Example: vmx-13 - This parameter may be omitted when using the VmSpec to update the contents of an existing VM.
+	VmToolsVersion    string            `xml:"VmToolsVersion,omitempty"`    // VMware tools version of this VM.
+	VirtualCpuType    string            `xml:"VirtualCpuType,omitempty"`    // The capabilities settings for this VM. This parameter may be omitted when using the VmSpec to update the contents of an existing VM.
+	TimeSyncWithHost  *bool             `xml:"TimeSyncWithHost,omitempty"`  // Synchronize the VM's time with the host.
+}
+
+type RecomposeVAppParamsForEmptyVm struct {
```
Comment (Didainius) [inline types/v56/vm_types.go]: is there any difference in the fields used by `ReComposeVAppParams`

File: types/v56/vm_types.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	XMLName          xml.Name    `xml:"RecomposeVAppParams"`
+	XmlnsVcloud      string      `xml:"xmlns,attr"`
+	XmlnsOvf         string      `xml:"xmlns:ovf,attr"`
+	CreateItem       *CreateItem `xml:"CreateItem,omitempty"`
+	AllEULAsAccepted bool        `xml:"AllEULAsAccepted,omitempty"`
+}
+
+type CreateItem struct {
```
Comment (Didainius) [inline types/v56/vm_types.go]: Missing description and I'd think the name is a bit too generic. What kind of item?

File: govcd/vm.go [hunk 4]
```diff
@@ -1355,3 +1355,33 @@ func (vm *VM) UpdateInternalDisksAsync(disksSettingToUpdate *types.VmSpecSection
 		}, vm.client.GetSpecificApiVersionOnCondition(">= 32.0", "32.0"))
 
 }
+
+func (vapp *VApp) AddEmptyVm(reComposeVAppParams *types.RecomposeVAppParamsForEmptyVm) (*VM, error) {
+	// TODO add validation
```
Comment (Didainius) [inline govcd/vm.go]: Is it self note or note for future?

File: types/v56/vm_types.go (hunk 3 above)
Comment (vbauzysvmware) [reply types/v56/vm_types.go]: it's XML element name, will add comment

File: types/v56/vm_types.go (hunk 2 above)
Comment (vbauzysvmware) [reply types/v56/vm_types.go]: big difference 
image

File: govcd/media.go (hunk 1 above)
Comment (vbauzysvmware) [reply govcd/media.go]: I think you query simple and type admin type as sys admin

File: govcd/vm.go (hunk 4 above)
Comment (vbauzysvmware) [reply govcd/vm.go]: implemented

File: types/v56/vm_types.go (hunk 2 above)
Comment (Didainius) [reply types/v56/vm_types.go:23]: My point was that RecomposeVAppParamsForEmptyVm is probably a subset of `RecomposeVAppParams`. That way we could reuse the same. But let it be

File: govcd/media.go (hunk 1 above)
Comment (Didainius) [reply govcd/media.go:649]: My point would be that name `QueryAllMedia` is misleading if an admin can query both, but would only get results from `adminMedia`. Although I don't know if this is a true statmenet

Comment (vbauzysvmware) [reply govcd/media.go:649]: My name intention was AllMedia - not bound by catalog.

File: types/v56/vm_types.go (hunk 2 above)
Comment (vbauzysvmware) [reply types/v56/vm_types.go:23]: It not always strgght forward to inherit like it is needed as XML has to comply exact Order

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 * Added methods `OrgVdcNetwork.Update`, `OrgVdcNetwork.UpdateAsync`, and `OrgVdcNetwork.Rename` [#292](https://github.com/vmware/go-vcloud-director/pull/292)
 * Added methods `EdgeGateway.Update` and `EdgeGateway.UpdateAsync` [#292](https://github.com/vmware/go-vcloud-director/pull/292)
+* Added methods `vapp.AddEmptyVm`, `vapp.AddEmptyVmAsync` and vdc.QueryAllMedia [#296](https://github.com/vmware/go-vcloud-director/pull/296)
```
Comment (dataclouder) [inline CHANGELOG.md:5]:
```suggestion
* Added methods `vapp.AddEmptyVm`, `vapp.AddEmptyVmAsync` and `vdc.QueryAllMedia` [#296](https://github.com/vmware/go-vcloud-director/pull/296)
```

File: govcd/api.go [hunk 6]
```diff
@@ -564,3 +564,8 @@ func combinedTaskErrorMessage(task *types.Task, err error) string {
 func takeBoolPointer(value bool) *bool {
 	return &value
 }
+
+// takeIntAddress is a helper which can gives address of `int`
```
Comment (dataclouder) [inline govcd/api.go:568]:
```suggestion
// takeIntAddress is a helper that returns the address of an `int`
```

File: govcd/media.go [hunk 7]
```diff
@@ -637,3 +637,42 @@ func (mediaRecord *MediaRecord) Delete() (Task, error) {
 	return mediaRecord.client.ExecuteTaskRequest(mediaRecord.MediaRecord.HREF, http.MethodDelete,
 		"", "error deleting Media item: %s", nil)
 }
+
+// QueryAllMedia returns media image found in system using `name` as query.
```
Comment (dataclouder) [inline govcd/media.go:641]:
```suggestion
// QueryAllMedia returns all media images found in system using `name` as query.
```

File: govcd/media_test.go [hunk 8]
```diff
@@ -333,3 +333,37 @@ func (vcd *TestVCD) Test_RefreshMediaRecord(check *C) {
...
+		return
+	}
+	// Fetching organization
+	org, err := vcd.client.GetAdminOrgByName(vcd.org.Org.Name)
+	check.Assert(err, IsNil)
+	check.Assert(org, NotNil)
+
+	testQueryMediaName := vcd.config.Media.Media
+
+	mediaSlise, err := vcd.vdc.QueryAllMedia(testQueryMediaName)
```
Comment (dataclouder) [inline govcd/media_test.go:358]:
```suggestion
	mediaList, err := vcd.vdc.QueryAllMedia(testQueryMediaName)
```

File: govcd/vm.go [hunk 9]
```diff
@@ -1352,6 +1352,126 @@ func (vm *VM) UpdateInternalDisksAsync(disksSettingToUpdate *types.VmSpecSection
 			Ovf:           types.XMLNamespaceOVF,
 			Name:          vm.VM.Name,
 			VmSpecSection: disksSettingToUpdate,
+			// API version requirements changes through vCD version to access VmSpecSection
 		}, vm.client.GetSpecificApiVersionOnCondition(">= 32.0", "32.0"))
 
 }
+
+// AddEmptyVm adds new empty VM (without template) to vApp and returns new created VM or error.
```
Comment (dataclouder) [inline govcd/vm.go:1360]:
```suggestion
// AddEmptyVm adds an empty VM (without template) to vApp and returns the new created VM or an error.
```

File: govcd/vm.go [hunk 10]
```diff
@@ -1352,6 +1352,126 @@ func (vm *VM) UpdateInternalDisksAsync(disksSettingToUpdate *types.VmSpecSection
...
+	newVm, err := vapp.GetVMByName(reComposeVAppParams.CreateItem.Name, true)
+	if err != nil {
+		return nil, err
+	}
+
+	return newVm, nil
+
+}
+
+// AddEmptyVmAsync adds new empty VM (without template) to vApp and returns Task or error.
```
Comment (dataclouder) [inline govcd/vm.go:1381]:
```suggestion
// AddEmptyVmAsync adds an empty VM (without template) to the vApp and returns a Task and an error.
```

File: govcd/vm.go [hunk 11]
```diff
@@ -1352,6 +1352,126 @@ func (vm *VM) UpdateInternalDisksAsync(disksSettingToUpdate *types.VmSpecSection
...
+
+	reComposeVAppParams.XmlnsVcloud = types.XMLNamespaceVCloud
+	reComposeVAppParams.XmlnsOvf = types.XMLNamespaceOVF
+
+	// Return the task
+	return vapp.client.ExecuteTaskRequest(apiEndpoint.String(), http.MethodPost,
+		types.MimeRecomposeVappParams, "error instantiating a new VM: %s", reComposeVAppParams)
+}
+
+func validateEmptyVmParams(reComposeVAppParams *types.RecomposeVAppParamsForEmptyVm) error {
```
Comment (dataclouder) [inline govcd/vm.go:1398]: Please add description

File: govcd/vm_test.go [hunk 12]
```diff
@@ -1296,3 +1296,186 @@ func deleteVapp(vcd *TestVCD, name string) error {
...
+	} else {
+		fmt.Println("Skipping adding another vdc network as network2 was not specified")
+	}
+
+	cat, err := vcd.org.GetCatalogByName(vcd.config.VCD.Catalog.Name, true)
+	check.Assert(err, IsNil)
+	check.Assert(cat, NotNil)
+
+	//media, err := cat.GetMediaByName("photon-custom-hw11-2.0-304b817.ova", false)
+	media, err := cat.GetMediaByName("vaido2", false)
```
Comment (dataclouder) [inline govcd/vm_test.go:1357]: hardcoded name. How does this media item become available?

File: types/v56/vm_types.go [hunk 13]
```diff
@@ -0,0 +1,39 @@
+package types
```
Comment (dataclouder) [inline types/v56/vm_types.go:1]: Why a separate file for some VM related types?
If we split by entity, this file should contain all VM types, but then we should have separate files for each entity

File: CHANGELOG.md (hunk 5 above)
Comment (vbauzysvmware) [reply CHANGELOG.md]: updated

File: govcd/api.go (hunk 6 above)
Comment (vbauzysvmware) [reply govcd/api.go]: updated

File: govcd/media.go (hunk 7 above)
Comment (vbauzysvmware) [reply govcd/media.go]: updated

File: govcd/media_test.go (hunk 8 above)
Comment (vbauzysvmware) [reply govcd/media_test.go]: updated

File: govcd/vm.go (hunk 9 above)
Comment (vbauzysvmware) [reply govcd/vm.go]: updated

File: govcd/vm.go (hunk 10 above)
Comment (vbauzysvmware) [reply govcd/vm.go]: updated

File: govcd/vm.go (hunk 11 above)
Comment (vbauzysvmware) [reply govcd/vm.go:1399]: added

File: govcd/vm_test.go (hunk 12 above)
Comment (vbauzysvmware) [reply govcd/vm_test.go]: updated. Value comes from configuration

File: types/v56/vm_types.go (hunk 13 above)
Comment (vbauzysvmware) [reply types/v56/vm_types.go:1]: I am trying to start splitting by domain(VM/VAPP/EdgeGtw/ etc.). If everyone agrees then it is start to add new one to separate `type` files. In time we move everything accordingly. I won't do let's split everything now it's to much and not this PR scope, but we can start doing better from today.

**Pull request merged**